File: apps/desktop/src/features/session/mountQueryBridge/series.ts
import type { MountId, PrSeriesId, PrSeriesView, ProjectId, SessionId } from '@goodboy/types';
import { resolveParentRequest } from '../../../store/slices/pr-series/parentRequest';
import { useAppStore } from '../../../store/store';

type BridgeArgs = Readonly<Record<string, unknown>>;

export type SeriesBridgeRequest = {
  readonly id: string;
  readonly provider: 'series';
  readonly verb: string;
  readonly sessionId: SessionId;
  readonly projectId: ProjectId | null;
  readonly mountId: MountId | null;
  readonly requestId?: string;
  readonly args: BridgeArgs;
};

export type SeriesBridgeOutcome = {
  readonly ok: boolean;
  readonly error?: string;
  readonly code?: string;
  readonly data?: unknown;
};

const text = ({ args, key }: { readonly args: BridgeArgs; readonly key: string }): string => {
  const value = args[key];
  return typeof value === 'string' ? value.trim() : '';
};

const count = ({
  args,
  key,
}: {
  readonly args: BridgeArgs;
  readonly key: string;
}): number | null => {
  const value = args[key];
  return typeof value === 'number' ? value : null;
};

const truthy = ({ args, key }: { readonly args: BridgeArgs; readonly key: string }): boolean =>
  args[key] === true;

const seriesResult = (view: PrSeriesView): Record<string, unknown> => ({
  seriesId: view.id,
  projectId: view.projectId,
  name: view.name,
  plannedCount: view.plannedCount,
  workItem: view.workItemIdentifier,
  workItemUrl: view.workItemUrl,
  parentRequest: view.parentRequest,
  members: view.members.map((member) => ({
    memberId: member.id,
    position: member.ordinal,
    label: member.label,
    status: member.status,
    mountId: member.mountId,
    branch: member.branch,
    request:
      member.request === null
        ? null
        : {
            provider: member.request.provider,
            host: member.request.host,
            repo: member.request.repoSlug,
            number: member.request.prNumber,
            url: member.request.url,
            state: member.request.state,
          },
  })),
});

type Params = {
  readonly request: SeriesBridgeRequest;
};

const create = async ({ request }: Params): Promise<SeriesBridgeOutcome> => {
  if (request.projectId === null) {
    return {
      ok: false,
      error: 'a series belongs to one project: name it with --project <name>',
      code: 'mount_unavailable',
    };
  }
  const parentRequest = resolveParentRequest({
    provider: text({ args: request.args, key: 'parentProvider' }),
    host: text({ args: request.args, key: 'parentHost' }),
    repo: text({ args: request.args, key: 'parentRepo' }),
    number: count({ args: request.args, key: 'parentNumber' }),
  });
  const workItem = text({ args: request.args, key: 'workItem' });
  const workItemUrl = text({ args: request.args, key: 'workItemUrl' });
  const series = await useAppStore.getState().createPrSeries({
    sessionId: request.sessionId,
    projectId: request.projectId,
    name: text({ args: request.args, key: 'name' }),
    plannedCount: count({ args: request.args, key: 'total' }),
    workItemIdentifier: workItem === '' ? null : workItem,
    workItemUrl: workItemUrl === '' ? null : workItemUrl,
    parentRequest,
  });
  return {
    ok: true,
    data: {
      seriesId: series.id,
      projectId: series.projectId,
      name: series.name,
      plannedCount: series.plannedCount,
    },
  };
};

const setMember = async ({ request }: Params): Promise<SeriesBridgeOutcome> => {
  const position = count({ args: request.args, key: 'position' });
  if (position === null) {
    return { ok: false, error: '--position needs a positive whole number' };
  }
  const label = text({ args: request.args, key: 'label' });
  const member = await useAppStore.getState().setPrSeriesMember({
    sessionId: request.sessionId,
    seriesId: text({ args: request.args, key: 'series' }) as PrSeriesId,
    position,
    mountId: request.mountId,
    label: label === '' ? null : label,
    isOmitted: truthy({ args: request.args, key: 'omitted' }),
  });
  return {
    ok: true,
    data: {
      seriesId: member.seriesId,
      memberId: member.id,
      position: member.ordinal,
      mountId: member.mountId,
      branch: member.branch,
      status: member.status,
    },
  };
};

const list = async ({ request }: Params): Promise<SeriesBridgeOutcome> => {
  const views = await useAppStore.getState().loadPrSeries({
    sessionId: request.sessionId,
    ...(request.projectId === null ? {} : { projectId: request.projectId }),
  });
  return { ok: true, data: { series: views.map(seriesResult) } };
};

export const executeSeriesRequest = async ({ request }: Params): Promise<SeriesBridgeOutcome> => {
  switch (request.verb) {
    case 'create':
      return await create({ request });
    case 'set-member':
      return await setMember({ request });
    case 'list':
      return await list({ request });
    default:
      return { ok: false, error: `unhandled series command: ${request.verb}` };
  }
};

File: apps/desktop/src-tauri/src/query_bridge/series.rs
use serde_json::{json, Value};
use tauri::AppHandle;

use super::dispatch::Scope;
use super::mount::handoff;
use super::protocol::{BridgeError, MOUNT_UNAVAILABLE};

const NO_SESSION: &str = "no session context: this command only works inside a Goodboy agent turn";
const NO_PROJECT: &str = "a series belongs to one project: name it with --project <name>";

fn text(args: &super::dispatch::Args, key: &str) -> Result<String, BridgeError> {
    args.get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
        .ok_or_else(|| format!("--{} must not be empty", key).into())
}

fn optional_text(args: &super::dispatch::Args, key: &str) -> Option<String> {
    args.get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
}

fn optional_number(args: &super::dispatch::Args, key: &str) -> Option<i64> {
    args.get(key).and_then(Value::as_i64)
}

fn flag(args: &super::dispatch::Args, key: &str) -> bool {
    args.get(key).and_then(Value::as_bool).unwrap_or(false)
}

fn required_number(args: &super::dispatch::Args, key: &str) -> Result<i64, BridgeError> {
    let Some(value) = optional_number(args, key) else {
        return Err(format!("--{} needs a whole number", key).into());
    };
    if value <= 0 {
        return Err(format!("--{} must be a positive whole number", key).into());
    }
    Ok(value)
}

pub(super) async fn dispatch(
    app: &AppHandle,
    scope: &Scope<'_>,
    verb: &str,
) -> Result<Value, BridgeError> {
    if scope.session.is_empty() {
        return Err(NO_SESSION.into());
    }
    let args = scope.args;
    match verb {
        "create" => {
            let Some(project) = scope.project_id() else {
                return Err(BridgeError::coded(MOUNT_UNAVAILABLE, NO_PROJECT));
            };
            let request_id = text(args, "request-id")?;
            handoff(
                app,
                json!({
                    "provider": "series",
                    "verb": "create",
                    "sessionId": scope.session,
                    "workspaceId": scope.workspace,
                    "projectId": project,
                    "mountId": scope.mount_id(),
                    "requestId": request_id,
                    "args": {
                        "name": text(args, "name")?,
                        "total": optional_number(args, "total"),
                        "workItem": optional_text(args, "work-item"),
                        "workItemUrl": optional_text(args, "work-item-url"),
                        "parentProvider": optional_text(args, "parent-provider"),
                        "parentHost": optional_text(args, "parent-host"),
                        "parentRepo": optional_text(args, "parent-repo"),
                        "parentNumber": optional_number(args, "parent-number"),
                    },
                }),
            )
            .await
        }
        "set-member" => {
            let request_id = text(args, "request-id")?;
            handoff(
                app,
                json!({
                    "provider": "series",
                    "verb": "set-member",
                    "sessionId": scope.session,
                    "workspaceId": scope.workspace,
                    "projectId": scope.project_id(),
                    "mountId": scope.mount_id(),
                    "requestId": request_id,
                    "args": {
                        "series": text(args, "series")?,
                        "position": required_number(args, "position")?,
                        "label": optional_text(args, "label"),
                        "omitted": flag(args, "omitted"),
                    },
                }),
            )
            .await
        }
        "list" => {
            handoff(
                app,
                json!({
                    "provider": "series",
                    "verb": "list",
                    "sessionId": scope.session,
                    "workspaceId": scope.workspace,
                    "projectId": scope.project_id(),
                    "mountId": scope.mount_id(),
                    "args": {},
                }),
            )
            .await
        }
        other => Err(format!("unhandled series command: {}", other).into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeMap;

    fn args(pairs: &[(&str, Value)]) -> super::super::dispatch::Args {
        let mut map: BTreeMap<String, Value> = BTreeMap::new();
        for (key, value) in pairs {
            map.insert((*key).to_string(), value.clone());
        }
        map
    }

    #[test]
    fn a_position_below_one_is_refused_before_the_app_is_asked() {
        let error = required_number(&args(&[("position", Value::from(0))]), "position")
            .expect_err("zero is not a position");

        assert!(error.message.contains("positive"));
    }

    #[test]
    fn a_missing_position_names_its_own_flag() {
        let error = required_number(&args(&[]), "position").expect_err("no position");

        assert!(error.message.contains("--position"));
    }

    #[test]
    fn an_absent_optional_parent_number_stays_absent() {
        assert_eq!(optional_number(&args(&[]), "parent-number"), None);
        assert_eq!(
            optional_number(
                &args(&[("parent-number", Value::from(198))]),
                "parent-number"
            ),
            Some(198)
        );
    }
}
